elastic/kibana
# [Console] Files and File Tree

## Summary

Introduce multiple text buffers (or files) into Console.

## Ship list

- [x] Implement a way to delete files
- [x] Implement a way to edit file names
- ~Investigate possibility of including directories (may nullify need for ^)~ Skipping dirs for now
- [x] Implement a way to search/filter files by name
- [x] Write tests for `editor` store reducer
- [x] test a11y
- [x] Review styling
- [x] Update screenshots!
- [x] Handle error case for when saving failed (surface in UI)
- [x] Alert the user to any unsaved state before the leave Console (not _super_ important because everything is still saved in localStorage)
- [x] Copy improvements (when there is an error for example, we should probably add a tooltip to tell the user what went wrong)
- [ ] Add a basic functional test for creating a new file, editing it and switching back to the initial file.

## Bonus features

- [ ] Remember last scroll position in file
- [ ] Implement a way to update file sorting: name (asc or desc), date created (asc or desc)
- ~Add directories - might form part of MVP~ This will form part of future work

## How to test

Start Kibana and navigate to Console. There should be an additional menu item `Files` that can be used to create, select and edit files.

- Try creating files
- Switching between files
- Interacting with the editor (typing text with the file tree open and closed).
- Open History + File Tree and watch out for any UI weirdness. The different menus in Console should still interact in a way that doesn't break anything.


After testing this branch, it would be good to check out master, go to Console and ensure that your current Console state is not lost. Data migration to text object format should already have occurred and so your current master state should be unaffected.

## FAQs

**How does this interact with `load_from` behaviour?**

The `load_from` behaviour (typically used inside of docs as a link called "View in Console") will ~still overwrite the currently loaded file content as it did before~ overwrite the scratchpad as this is opened by default.

**What is ~the scratch pad~  _Default_?**

The _Default_ text buffer is intended as the starting point in Console. The _Default_ text buffer also persists its content but unlike other text objects cannot be deleted.

**What about directories?**

Currently there are no directories. This still needs to be investigated further, but is not currently part of the MVP. Subject to change.

**Can I export files?**

No. This is will be a future addition.

**What about `LocalStorage` + `SessionStorage` storage limits?**

Arguably, with multiple files, there is an increased risk of hitting storage size limits (very large mappings objects can be up to 300KB of text). According to [wikipedia](https://en.wikipedia.org/wiki/Web_storage) localStorage storage quota is 10MB so users can store ~33 requests with _very_ large mappings (usually ~10 000 lines long each).

At present, the code does not "know" about localStorage but interacts with a storage adapter. If we encounter an error like this we should surface it in the UI (like any storage error), guiding users to a solution. Also, this error will not stop users from typing new text, it will just not be persisted.

Having that amount of text is very unwieldly and it is expected that only a small subset of users will encounter this issue (if any).

## Screenshots

<details>
<summary>Different file tree states</summary>

<img width="1255" alt="Screenshot 2020-03-10 at 09 26 54" src="https://user-images.githubusercontent.com/8155004/76297634-71bdf380-62b8-11ea-9c20-1110c4def190.png">

<img width="1256" alt="Screenshot 2020-03-10 at 09 29 33" src="https://user-images.githubusercontent.com/8155004/76297636-74204d80-62b8-11ea-94cd-b7a5f5cb0255.png">

<img width="1258" alt="Screenshot 2020-03-10 at 09 36 13" src="https://user-images.githubusercontent.com/8155004/76297639-75517a80-62b8-11ea-8f5c-bea0e22211c9.png">

</details>

### Checklist

Delete any items that are not applicable to this PR.


### For maintainers
